### PR TITLE
[FLINK-26235][connectors/filesystem] CompactingFileWriter and PendingFileRecoverable should not be exposed to users.

### PR DESCRIPTION
**Pull request by pltbkd**
## What is the purpose of the change

In [FLINK-25583](https://issues.apache.org/jira/browse/FLINK-25583), we added a FileCompactingWriter to write the compacting file. The writer is exposed to users in the FileCompactor, while it has a `closeForCommit` function, and relies on the `PendingFileRecoverable`, both of which should not be exposed to users. We should fix this and annotate them as Internal.

## Brief change log

  - Annotate CompactingFileWriter and PendingFileRecoverable as Internal.
  - Hide CompactingFileWriter in the FileCompactor.
  - Call FileCompactor with either of the implementation.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: docs/content/docs/connectors/datastream/filesystem.md [hunk 1]
```diff
@@ -1016,10 +1016,14 @@ Once the total size of the cached files has reached the size threshold or the nu
...
+the give list of `Path` and write the result file. It could be classified into two types according to how to write the file:
+
+- **{{< javadoc file="org/apache/flink/connector/file/sink/compactor/OutputStreamBasedFileCompactor.html" name="OutputStreamBasedFileCompactor">}}**: 
+  The users can write the compacted results into an output stream. This is useful when the users don't want to or can't read records from the input files. 
+  An example is the {{< javadoc file="org/apache/flink/connector/file/sink/compactor/ConcatFileCompactor.html" name="ConcatFileCompactor">}} that concats the list of files directly.
+- **{{< javadoc file="org/apache/flink/connector/file/sink/compactor/RecordWiseFileCompactor.html" name="RecordWiseFileCompactor">}}**: 
+  The compactor can read records one-by-one from the input files and write into the result file as like the `FileWriter`.
```
Comment (gaoyunhaii) [inline docs/content/docs/connectors/datastream/filesystem.md:1025]: `as like the` -> `similar to` ?

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactService.java [hunk 2]
```diff
@@ -95,11 +102,22 @@ public void close() {
...
         CompactingFileWriter compactingFileWriter =
                 bucketWriter.openNewCompactingFile(
-                        fileCompactor.getWriterType(),
+                        compactingWriterType,
                         request.getBucketId(),
                         targetPath,
                         System.currentTimeMillis());
-        fileCompactor.compact(compactingFiles, compactingFileWriter);
+        if (compactingWriterType == Type.RECORD_WISE) {
+            ((RecordWiseFileCompactor) fileCompactor)
```
Comment (gaoyunhaii) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactService.java:110]: Might change to `RecordWiseFileCompactor<?>`

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactService.java [hunk 3]
```diff
@@ -95,11 +102,22 @@ public void close() {
...
+                    .compact(
+                            compactingFiles,
+                            ((RecordWiseCompactingFileWriter) compactingFileWriter)::write);
```
Comment (gaoyunhaii) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactService.java:113]: Might add `@SuppressWarnings({"rawtypes", "unchecked"})` to the method.

**Review by gaoyunhaii — APPROVED** [review]
Thanks @pltbkd for the fix! LGTM and I only have small comments.

**Comment by pltbkd** [comment]
@gaoyunhaii 
Thanks for the review! The comments are now resolved.

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Pull request closed without merging**